R/stream.R

```diff
@@ -1,443 +1,230 @@
-#' x <- lat_lng(rtworld)
+#' # Download another 10s worth of data to the same file
+#' e <- stream_tweets("election", timeout = 10)
 #'
+#' # stream tweets about continential USA for 5 minutes
+#' usa <- stream_tweets(lookup_coords("usa"), file_name = "usa.json", timeout = 300)
+#' 
 #' }
-#'
-#' @return Tweets data returned as data frame with users data as attribute.
-#' @family stream tweets
+#' @return A tibble with one row per tweet
 #' @export
 stream_tweets <- function(q = "",
                           timeout = 30,
                           parse = TRUE,
                           token = NULL,
                           file_name = NULL,
                           verbose = TRUE,
+                          append = TRUE,
                           ...) {
-  if ("append" %in% names(list(...))) {
-    stop("append should only be used with stream_tweets2() (which is in development)",
-         call. = FALSE)
-  }
-  ## set encoding
-  if (!identical(getOption("encoding"), "UTF-8")) {
-    op <- getOption("encoding")
-    options(encoding = "UTF-8")
-    on.exit(options(encoding = op), add = TRUE)
-  }
-  token <- check_token(token)
-  if (!timeout) {
-    timeout <- Inf
-  }
-  ## setp q to blank for sample stream
-  if (missing(q) || is.null(q)) q <- ""
-  stopifnot(
-    is.numeric(timeout),
-    timeout > 0,
-    any(is.atomic(q), inherits(q, "coords")),
-    is.atomic(file_name)
-  )
-  ## select stream API
-  if (identical(q, "")) {
-    query <- "statuses/sample"
-    params <- NULL
-  } else {
-    query <- "statuses/filter"
-    params <- stream_params(q, ...)
-  }
-  url <- make_url(
-    restapi = FALSE,
-    query,
-    param = params
-  )
-  ## temp file if none provided
   if (is.null(file_name)) {
-    tmp <- TRUE
-    file_name <- tmp_json()
+    file_name <- "stream_tweets.json"
+    inform(paste0("Writing to '", file_name, "'"))
+  }
+  output <- file(file_name)
+  
+  prep <- stream_prep(token, q, ...)
+  stream <- curl::curl(prep$url, handle = prep$handle)
+
+  quiet_interrupt(download_from_stream(stream, output, 
+    timeout = timeout,
+    verbose = verbose
+  ))
+  
+  if (parse) {
+    df <- jsonlite::stream_in(file(file_name), verbose = FALSE)
+    tibble::as_tibble(df)
   } else {
-    tmp <- FALSE
+    NULL
   }
-  if (is.infinite(timeout)) tmp <- FALSE
-  if (!grepl("\\.json$", file_name)) {
-    file_name <- paste0(file_name, ".json")
+}
+
+download_from_stream <- function(stream, output, append = TRUE, timeout = 10, verbose = TRUE) {
+  if (!timeout) {
+    timeout <- Inf
   }
-  if (!file.exists(file_name)) file.create(file_name)
+  stopifnot(is.numeric(timeout), timeout > 0)
+  stop_time <- Sys.time() + timeout
+  
+  n_seen <- 0
   if (verbose) {
-    message(
-      paste0("Streaming tweets for ", timeout, " seconds...")
+    pb <- progress::progress_bar$new(
+      total = NA,
+      show_after = 0,
+      format = "Streaming tweets: :n tweets written / :rate / :elapsedfull"
```

Comment:
Adding a \n at the end of the message will help with the display that @jthomasmock comments. But will create a new message every 10 seconds or so.

Reply:
I can just add a separate counter for number of bytes.